File: leetcode/142. 环形链表 II.go
package main

// 142. 环形链表 II

func main() {

}

// 1. 遍历所有节点，使用map记录已经访问过的节点
// 2. 快慢指针 -- 两指针相遇后，再定义一个慢指针从链表头出发，当快慢指针再次相遇时，它指向的就是环的入口
func detectCycle(head *ListNode) *ListNode {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next

		if slow == fast {
			slow = head
			for slow != fast {
				slow = slow.Next
				fast = fast.Next
			}
			return slow
		}
	}
	return nil
}

File: C++/142.环形链表II.cpp

// 142. 环形链表 II
//
// 给定一个链表的头节点 head ，返回链表开始入环的第一个节点。 如果链表无环，则返回 null。

#include <header.h>

class Solution {
public:
    // 141
    // 使用快慢指针确认链表有环后, 两指针一个从head开始一个从相遇位置开始,
    // 继续向前移动, 直到相遇时相遇点就是入环的起点
    ListNode* detectCycle(ListNode* head) {
        ListNode *fast = head, *slow = head;
        while (fast != nullptr && fast->next != nullptr) {
            fast = fast->next->next;
            slow = slow->next;

            if (fast == slow) {
                fast = head;
                while (fast != slow) {
                    fast = fast->next;
                    slow = slow->next;
                }
                return slow;
            }
        }
        return nullptr;
    }
};
